### PR TITLE
Qtifw build installer

### PR DESCRIPTION
**Pull request by jroweboy**
Depends on: https://github.com/citra-emu/ext-windows-bin/pull/10

Supercedes: https://github.com/citra-emu/citra/pull/2929

(Got tired of waiting for @j-selby to get internet back)

This reworks the installer to be built through cmake instead of powershell/bash scripts. I tested windows, but if someone would be so kind as to test macOS and linux.

After this PR, there is one more followup PR for QT gui integration to allow users to update through the UI. When that follow up PR is merged, the download page will need to be reworked as well to point directly to the new installer instead. (We will need lots of testing of course!)

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/2966)

### CODE REVIEW

File: CMakeModules/BuildInstaller.cmake [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+file(MAKE_DIRECTORY ${DIST_DIR})
+file(DOWNLOAD https://github.com/citra-emu/ext-windows-bin/raw/master/qtifw/${ARCHIVE}
+    "${BUILD_DIR}/${ARCHIVE}" SHOW_PROGRESS)
+execute_process(COMMAND ${CMAKE_COMMAND} -E tar xf "${BUILD_DIR}/${ARCHIVE}"
+    WORKING_DIRECTORY "${BUILD_DIR}/")
+
+set(CONFIG_FILE "${SRC_DIR}/config/config_${PLATFORM}.xml")
+set(INSTALLER_BASE "${BUILD_DIR}/installerbase_${PLATFORM}")
+set(BINARY_CREATOR "${BUILD_DIR}/binarycreator_${PLATFORM}")
```
Comment (B3n30) [inline CMakeModules/BuildInstaller.cmake]: chmod +x needs to be added to binarycreator_${PLATFORM} (at least on osx)

File: CMakeModules/BuildInstaller.cmake [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+elseif(LINUX)
+    set(PLATFORM "linux")
+else()
+    message(FATAL_ERROR "Cannot build installer for this unsupported platform")
+endif()
+
+set(DIST_DIR "${BUILD_DIR}/dist")
+set(ARCHIVE "${PLATFORM}.7z")
+
+file(MAKE_DIRECTORY ${DIST_DIR})
```
Comment (B3n30) [inline CMakeModules/BuildInstaller.cmake]: file(MAKE_DIRECTORY ${BUILD_DIR})

File: CMakeModules/BuildInstaller.cmake [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+execute_process(COMMAND ${CMAKE_COMMAND} -E tar xf "${BUILD_DIR}/${ARCHIVE}"
+    WORKING_DIRECTORY "${BUILD_DIR}/")
+
+set(CONFIG_FILE "${SRC_DIR}/config/config_${PLATFORM}.xml")
+set(INSTALLER_BASE "${BUILD_DIR}/installerbase_${PLATFORM}")
+set(BINARY_CREATOR "${BUILD_DIR}/binarycreator_${PLATFORM}")
+set(PACKAGES_DIR "${BUILD_DIR}/packages")
+file(MAKE_DIRECTORY ${PACKAGES_DIR})
+
+execute_process(COMMAND ${BINARY_CREATOR} -t ${INSTALLER_BASE} -n -c ${CONFIG_FILE} -p ${PACKAGES_DIR} ${TARGET_FILE})
```
Comment (B3n30) [inline CMakeModules/BuildInstaller.cmake]: chmod +x ${TARGET_FILE}.app/Contents/MacOS/${TARGET_FILE}

File: dist/installer/README.md [hunk 4]
```diff
@@ -0,0 +1,9 @@
+Citra Qt Installer
+==================
+
+This contains the configuration files for building Citra's installer.
+
+Installers can only be built on the platform that they are targeting.
+
+Build the `installer` target to generate the installer, and the installer will be in
```
Comment (B3n30) [inline dist/installer/README.md]: Can be in an later PR but this should be better explained depending on the os

**Comment by jroweboy** [comment]
@B3n30 the installer is an online installer, meaning its only created once and it will read package information from the internet. This means that we won't need to do anything as part of the CI process since its "deploy once use forever (tm)" The only time we'll need to rebuild the installer, is when we upgrade to a newer version of the library (version 3 came out recently actually) and we can build it manually and upload it to the web server like we already have done. The point of this PR is just to make it easy for us to build it again in the future (and have some documentation about how to build it)

File: dist/installer/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+# Adds a custom target that will run the BuildInstaller.cmake file
+# CMake can't just run a cmake function as a custom command, so this is a way around it.
+# Calls the cmake command and runs a cmake file in "scripting" mode passing in variables with -D
+add_custom_command(OUTPUT "${TARGET_FILE}"
+    COMMAND ${CMAKE_COMMAND} -DSRC_DIR=${CMAKE_CURRENT_SOURCE_DIR} -D BUILD_DIR=${BUILD_DIR} -D TARGET_FILE=${TARGET_FILE} -P ${CMAKE_SOURCE_DIR}/CMakeModules/BuildInstaller.cmake
+    WORKING_DIRECTORY ${BUILD_DIR}
+)
+
+add_custom_target(installer DEPENDS ${TARGET_FILE})
```
Comment (j-selby) [inline dist/installer/CMakeLists.txt]: Newline

File: dist/installer/README.md (hunk 4 above)
Comment (j-selby) [reply dist/installer/README.md]: Don't think it needs to be overly verbose. Direct CMake instructions would be good though.

File: dist/installer/config/config_linux.xml [hunk 6]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<Installer>
+    <Name>Citra</Name>
+    <Version>1.0.0</Version>
+    <Title>Citra Updater</Title>
+    <Publisher>Citra team</Publisher>
+    <!-- e.g. /home/<user>/Citra -->
+    <TargetDir>@HomeDir@/Citra</TargetDir>
```
Comment (TheKoopaKingdom) [inline dist/installer/config/config_linux.xml]: I'm not sure if this behavior is really expected? Generally Linux binaries aren't stored in the home directory, in favor of `/usr/share/`.

Comment (j-selby) [reply dist/installer/config/config_linux.xml]: The issue here was that QtIFW cannot sudo up by itself (bugged), so it crashed when writing to those directories. To make its behaviour logical, we write to here.

File: dist/installer/README.md (hunk 4 above)
Comment (jroweboy) [reply dist/installer/README.md]: its different for every generator :P not sure i care enough to explain how to build a specific target

File: dist/installer/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+endif()
+
+set(BUILD_DIR "${CMAKE_BINARY_DIR}/installer")
+set(DIST_DIR "${BUILD_DIR}/dist")
+set(TARGET_FILE "${DIST_DIR}/citra-setup-${PLATFORM}")
+
+# Adds a custom target that will run the BuildInstaller.cmake file
+# CMake can't just run a cmake function as a custom command, so this is a way around it.
+# Calls the cmake command and runs a cmake file in "scripting" mode passing in variables with -D
+add_custom_command(OUTPUT "${TARGET_FILE}"
```
Comment (B3n30) [inline dist/installer/CMakeLists.txt]: the installer dir is needs to exist here. So the make_directory command needs to be executed before that command.

FYI thats the command it tries to execute but fails with the cd
`cd /Users/bthomas/citra/build/installer && /Applications/CMake.app/Contents/bin/cmake -DSRC_DIR=/Users/bthomas/citra/dist/installer -D BUILD_DIR=/Users/bthomas/citra/build/installer -D TARGET_FILE=/Users/bthomas/citra/build/installer/dist/citra-setup-mac -P /Users/bthomas/citra/CMakeModules/BuildInstaller.cmake`

File: dist/installer/config/config_linux.xml [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    <InstallerApplicationIcon>icon</InstallerApplicationIcon>
```
Comment (j-selby) [inline dist/installer/config/config_linux.xml]:
```xml
<TargetDir>@HomeDir@/Citra</TargetDir>
<AdminTargetDir>/opt/citra</AdminTargetDir>
```
maybe?

File: dist/installer/config/config_windows.xml [hunk 9]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<Installer>
+    <Name>Citra</Name>
+    <Version>1.0.0</Version>
+    <Title>Citra Updater</Title>
+    <Publisher>Citra team</Publisher>
+    <StartMenuDir>Citra</StartMenuDir>
+    <!-- e.g. C:\home\<user>\AppData\Local\Citra -->
+    <TargetDir>@HomeDir@/AppData/Local/Citra</TargetDir>
```
Comment (j-selby) [inline dist/installer/config/config_windows.xml]: Do we want this in `%appdata%` or `C:\Program Files`, now we have a mechanism to write there (with QtIFW 3)?

File: dist/installer/config/config_linux.xml (hunk 6 above)
Comment (jroweboy) [reply dist/installer/config/config_linux.xml]: We migrated to QTIFW 3.0 which resolves the issue. It also added an option for installing to a different directory if run as root. So now when installing without root, we'll put it in `~/.Citra` When installing with root, we'll put it in `/opt/Citra`

File: dist/installer/config/config_windows.xml [hunk 10]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<Installer>
+    <Name>Citra</Name>
+    <Version>1.0.0</Version>
+    <Title>Citra Updater</Title>
+    <Publisher>Citra team</Publisher>
```
[the remark this answers is not in the document]
Comment (jroweboy) [reply dist/installer/config/config_windows.xml]: check the latest commit

File: dist/installer/config/config_linux.xml (hunk 6 above)
Comment (chris062689) [reply dist/installer/config/config_linux.xml]: Do we prompt for root at the start of the installer?

Comment (jroweboy) [reply dist/installer/config/config_linux.xml]: no

**Comment by iMonZ** [comment]
Can this merge please be reverted.
It doesn't work great with brew on macOS, and a regular DMG or pkg would be enough!
Thanks!

**Pull request merged**